benchmarks-mobile/protonano_app/app/src/main/java/io/grpc/benchmarks/ProtobufBenchmarker.java

```diff
@@ -0,0 +1,165 @@
+            @Override
+            public void execute() throws Exception {
+                MessageNano.mergeFrom(message.getClass().newInstance(), inputData);
+            }
+        });
+    }
+
+    public static BenchmarkResult serializeJsonToByteArray(final String jsonString, boolean gzip)
+            throws Exception {
+        final int serializedSize = jsonString.getBytes().length;
+        final JSONObject jsonObject = new JSONObject(jsonString);
+
+        if (gzip) {
+            ByteArrayOutputStream bos = new ByteArrayOutputStream(serializedSize);
+            GZIPOutputStream gos = new GZIPOutputStream(bos);
+            gos.write(jsonString.getBytes());
+            gos.close();
+            bos.close();
+
+            BenchmarkResult res = benchmark("JSON serialize to byte array (gzip)", serializedSize,
+                    new Action() {
+                @Override
+                public void execute() throws IOException {
+                    ByteArrayOutputStream bos = new ByteArrayOutputStream(serializedSize);
+                    GZIPOutputStream gos = new GZIPOutputStream(bos);
+                    gos.write(jsonObject.toString().getBytes());
+                    gos.close();
+                    bos.close();
+                    bos.toByteArray();
+                }
+            });
+            res.compressedSize = bos.toByteArray().length;
+            return res;
+        } else {
+            return benchmark("JSON serialize to byte array", serializedSize, new Action() {
+                @Override
+                public void execute() throws JSONException {
+                    jsonObject.toString().getBytes();
+                }
+            });
+        }
+    }
+
+    public static BenchmarkResult deserializeJsonfromByteArray(final String jsonString,
+                                                               boolean gzip) throws Exception {
+        final int serializedSize = jsonString.getBytes().length;
+
+        if (gzip) {
+            ByteArrayOutputStream bos = new ByteArrayOutputStream(serializedSize);
+            GZIPOutputStream gos = new GZIPOutputStream(bos);
+            gos.write(jsonString.getBytes());
+            gos.close();
+            bos.close();
+            final byte[] compressedData = bos.toByteArray();
+
+            BenchmarkResult res = benchmark("JSON deserialize from byte array (gzip)",
+                    serializedSize, new Action() {
+                @Override
+                public void execute() throws JSONException, IOException {
+                    ByteArrayInputStream bis = new ByteArrayInputStream(compressedData);
+                    GZIPInputStream gis = new GZIPInputStream(bis, serializedSize);
+                    byte[] inputData = new byte[serializedSize];
+                    gis.read(inputData);
+                    gis.close();
+                    bis.close();
+                    new JSONObject(new String(inputData));
```

Comment:
Is JSONObject lazy?  Will it skip parsing most of the data until it is asked for?